src/services/withdrawal.py

```diff
@@ -0,0 +1,132 @@
+from web3.types import Wei
+
+from src.metrics.prometheus.business import CONTRACT_ON_PAUSE
+from src.variables import FINALIZATION_BATCH_MAX_REQUEST_COUNT
+from src.utils.abi import named_tuple_to_dataclass
+from src.web3py.typings import Web3
+from src.typings import ReferenceBlockStamp
+from src.services.safe_border import SafeBorder
+from src.modules.submodules.consensus import ChainConfig, FrameConfig
+from src.modules.accounting.typings import BatchState
+
+
+class Withdrawal:
+    """
+    Service calculates which withdrawal requests should be finalized using next factors:
+
+    1. Safe border epoch for the current reference slot.
+    2. The amount of available ETH is determined from the Withdrawal Vault, EL Vault, and buffered ETH.
+    """
+    def __init__(
+        self,
+        w3: Web3,
+        blockstamp: ReferenceBlockStamp,
+        chain_config: ChainConfig,
+        frame_config: FrameConfig,
+    ) -> None:
+        self.w3 = w3
+        self.safe_border_service = SafeBorder(self.w3, blockstamp, chain_config, frame_config)
+
+        self.chain_config = chain_config
+        self.frame_config = frame_config
+        self.blockstamp = blockstamp
+
+    def get_finalization_batches(
+        self,
+        is_bunker_mode: bool,
+        share_rate: int,
+        withdrawal_vault_balance: Wei,
+        el_rewards_vault_balance: Wei
+    ) -> list[int]:
+        on_pause = self._is_requests_finalization_paused()
+        CONTRACT_ON_PAUSE.set(on_pause)
+        if on_pause:
+            return []
+
+        if not self._has_unfinalized_requests():
+            return []
+
+        withdrawable_until_epoch = self.safe_border_service.get_safe_border_epoch(is_bunker_mode)
+        withdrawable_until_timestamp = (
```

Comment:
Recheck all condition and contract functions.
check Sanity checks and check calculateFinalizationBatches logic